lib/eslint.js

```diff
@@ -807,20 +811,18 @@ module.exports = (function() {
              */
             controller.traverse(ast, {
                 enter: function(node, parent) {
-
                     var comments = api.getComments(node);
 
                     emitCommentsEnter(comments.leading);
                     node.parent = parent;
-                    api.emit(node.type, node);
+                    codePathAnalyzer.enterNode(node);
```

Comment:
Ooh, option #3 is a really nice and clean way of doing it. It would also make further changes a lot easier.